### DISCUSSION TITLE
watchpoint on virtual memory space

### DISCUSSION THREAD

**Discussion by jotego** [mamedev/discussions#70] [Q&A]
I am looking into the Namco System 1 driver. It has a hack related to the MCU marked in the driver as a _very obscure point_ and I am trying to make sense of it by measuring the real MCU on the board.

Related to this, I am trying to see how the CPUs interact. But, I find it very hard to catch writes to memory because two of the CPUs use a memory mapper. This mapper has a _virtual_map_ address space (in namcos1.cpp)

```
void namcos1_state::virtual_map(address_map &map)
{
	map(0x2c0000, 0x2c1fff).w(FUNC(namcos1_state::_3dcs_w));
	map(0x2e0000, 0x2e7fff).rw(m_c116, FUNC(namco_c116_device::read), FUNC(namco_c116_device::write));
	map(0x2f0000, 0x2f7fff).rw(m_c123tmap, FUNC(namco_c123tmap_device::videoram8_r), FUNC(namco_c123tmap_device::videoram8_w));
	map(0x2f8000, 0x2f9fff).rw(FUNC(namcos1_state::no_key_r), FUNC(namcos1_state::no_key_w));
	map(0x2fc000, 0x2fcfff).ram().w(FUNC(namcos1_state::spriteram_w)).share("spriteram");
	map(0x2fd000, 0x2fd01f).rw(m_c123tmap, FUNC(namco_c123tmap_device::control8_r), FUNC(namco_c123tmap_device::control8_w)).mirror(0xfe0);
	map(0x2fe000, 0x2fe3ff).ram().rw("namco", FUNC(namco_cus30_device::namcos1_cus30_r), FUNC(namco_cus30_device::namcos1_cus30_w)).mirror(0xc00); /* PSG ( Shared ) */
	map(0x2ff000, 0x2ff7ff).ram().share("triram").mirror(0x800);
	map(0x300000, 0x307fff).ram();
	map(0x400000, 0x7fffff).rom().region("user1", 0);
}
```

It looks like I cannot normally add watchpoints to this memory map. I have to use the CPU regular memory map, which keeps changing as the CPU programs the mapper.

Is there a way to watch virtual memory maps in the debugger?

**Comment by cuavas** [comment]
The `virtual_map` is used by the C117 chip – you might be able to put a watchpoint on it like `wpset 1234:c117:program,1,rw`

You could also install taps on it with a Lua script.

**Comment by jotego** [comment]
Thank you. That sounds promising, although it doesn't work:

![imagen](https://github.com/mamedev/discussions/assets/1863036/9ce53d02-f464-4ac3-92b1-ac44661a32ac)

**Comment by cuavas** [comment]
Use Lua script (either with `-autoboot_script` or `-console`).  Something like:
```lua
vspace = manager.machine.devices['c117'].spaces['program']
action = function(offset, data) print("Watchpoint hit"); manager.machine.debugger.execution_state = 'stop'; end
wpr = vspace:install_read_tap(0x2fc000, 0x2fc000, "wp", action)
wpw = vspace:install_write_tap(0x2fc000, 0x2fc000, "wp", action)
```
